doc/developer/design/20220516_transactional_consistency.md

```diff
@@ -0,0 +1,244 @@
+and that the total order is constrained by the real time occurrences of those transactions. In other words, if
+transaction A finishes in real time before transaction B starts in real time, then transaction A MUST be ordered before
+transaction B in the total order. Concurrent transaction can be ordered arbitrarily. For a more in depth discussion
+please
+read [https://jepsen.io/consistency/models/strict-serializable](https://jepsen.io/consistency/models/strict-serializable)
+and chapters 7 and 9 of [Designing Data-Intensive Applications](https://dataintensive.net/).
+
+Materialize can satisfy the constraints of Strict Serializability if all reads and writes occur at specific timestamps,
+and these timestamps are non-decreasing in real-time order, with strict increases for writes following reads. Each
+timestamp is assigned to an event at some real-time moment within the event's bounds (its start and stop). Each
+timestamp transition is made durable before any subsequent response is issued.
+
+We can then use these timestamp to form a total order constrained by real time:
+
+- Transaction T1 is ordered before Transaction T2 if the timestamp of T1 < the timestamp of T2.
+- Transaction T1 is ordered before Transaction T2 if the timestamp of T1 == the timestamp of T2, T1 is write-only, and
+  T2 is read-only.
+- Transaction T1 can be arbitrarily ordered with Transaction T2 if the timestamp of T1 == the timestamp of T2, T1 is
+  read-only, and T2 is read-only.
+- Transaction T1 can be arbitrarily ordered with Transaction T2 if the timestamp of T1 == the timestamp of T2, T1 is
+  write-only, and T2 is write-only.
+
+### Global Timestamp
+
+All reads to any object and all writes to user tables are assigned a timestamp by the Coordinator on a single thread.
+The logic for determining what timestamp to return must satisfy:
+
+- If it's a read, then it is larger than or equal to the `since` of all objects involved in the query.
+- If it's a write, then it is larger than or equal to the `upper` of the table.
+- It is larger than or equal to the timestamp of the most recently completed query.
+    - If the query is a write and the most recently completed query was a read, then the timestamp returned must be
+      strictly larger than the previous timestamp.
+
+We can accomplish this by using a single global timestamp per timeline. We do not allow reads across timelines and
+therefore provide no consistency guarantees across timelines. For the `EpochMilliseconds` timeline, which is used for
+user tables and most sources, the global timestamp will be initialized to the current system time and be advanced
+periodically to the current value of a monotonically increasing system clock. All other global timestamps will be
+initialized to the timeline's minimum value and be periodically advanced in the following
+way: `global_timestamp = max(min(uppers) - 1, global_timestamp)`, where `uppers` is the list of all `upper`s in the
+timeline.
```

Comment:
I think this paragraph is too focused on how I think certain things should be implemented, rather than what's required to achieve strict serializability. 

Instead I think it needs to convey the following points:
- A read can only access objects from a single timeline.
- Each timeline needs to be initialized to some initial timestamp.
- Each timeline needs to be periodically advanced.
- The current timestamp of the timeline should be a valid read timestamp for all objects in the timeline.
- All reads are assigned a timestamp equal to the value of the timeline's timestamp.